### ISSUE TITLE
Deadlock after updating from 5.21.0 to 5.22.0

### ISSUE THREAD

**Issue by cosmer-work** [groue/GRDB.swift#1194]
### What did you do?

Updated from 5.21.0 to 5.22.0

### What did you expect to happen?

Nothing.

### What happened instead?

Crash due to deadlock.

### Environment

**GRDB flavor(s): GRDB
**GRDB version: 5.22.0
**Installation method: SPM
**Xcode version: 13.3
**Swift version: 5.6
**Platform(s) running GRDB: iOS
**macOS version running Xcode: 12.2.1

**Comment by cosmer-work** [comment]
Minimal reproducer.

```
struct Record: Codable, FetchableRecord, PersistableRecord {
    var id: Int64
}

var configuration = Configuration()
configuration.targetQueue = DispatchQueue(label: "crash.test", qos: .userInitiated)

let database = DatabaseQueue(configuration: configuration)

var migrator = DatabaseMigrator()
migrator.registerMigration("v1") { (db) in
    try db.create(table: Record.databaseTableName) { (t) in
        t.autoIncrementedPrimaryKey("id")
    }
}

try migrator.migrate(database)

let observation = ValueObservation.tracking { (db) in
    try Record.fetchCount(db)
}

var cancellable: AnyCancellable?
cancellable = observation.publisher(in: database, scheduling: .immediate)
    .map { _ in
        database.readPublisher { (db) in
            try Record.fetchCount(db)
        }
    }
    .switchToLatest()
    .sink { _ in

    } receiveValue: { (value) in
        print(value)
        cancellable?.cancel()
    }
```

**Comment by cosmer-work** [comment]
Thread 1 Queue : GRDB.DatabaseQueue (serial)
#0	0x0000000105fba58b in __DISPATCH_WAIT_FOR_QUEUE__ ()
#1	0x0000000105fb9f3a in _dispatch_sync_f_slow ()
#2	0x00007fff5967b3c4 in implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<τ_0_0>(execute:) ()
#3	0x00007fff5967ba92 in partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<τ_0_0>(execute:) ()
#4	0x00007fff5967ab1c in OS_dispatch_queue._syncHelper<τ_0_0>(fn:execute:rescue:) ()
#5	0x00007fff5967b06b in OS_dispatch_queue.sync<τ_0_0>(execute:) ()
#6	0x000000010584c504 in closure #1 in ValueWriteOnlyObserver.syncStart(from:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueWriteOnlyObserver.swift:204
#7	0x000000010585698f in partial apply for closure #1 in ValueWriteOnlyObserver.syncStart(from:) ()
#8	0x000000010562be7a in closure #1 in SerializedDatabase.reentrantSync<τ_0_0>(_:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/Core/SerializedDatabase.swift:160
#9	0x000000010562bf60 in partial apply for closure #1 in SerializedDatabase.reentrantSync<τ_0_0>(_:) ()
#10	0x00007fff5967c361 in partial apply for thunk for @callee_guaranteed () -> (@out τ_0_0, @error @owned Error) ()
#11	0x00007fff5967c5d9 in thunk for @callee_guaranteed () -> (@out τ_0_0, @error @owned Error)partial apply ()
#12	0x00007fff5967ac36 in closure #1 in closure #1 in OS_dispatch_queue._syncHelper<τ_0_0>(fn:execute:rescue:) ()
#13	0x00007fff5967c201 in partial apply for thunk for @callee_guaranteed () -> () ()
#14	0x00007fff5967aa0e in thunk for @escaping @callee_guaranteed () -> () ()
#15	0x0000000105faaa5b in _dispatch_client_callout ()
#16	0x0000000105fba666 in _dispatch_sync_invoke_and_complete_recurse ()
#17	0x00007fff5967b3c4 in implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<τ_0_0>(execute:) ()
#18	0x00007fff5967ba92 in partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<τ_0_0>(execute:) ()
#19	0x00007fff5967ab1c in OS_dispatch_queue._syncHelper<τ_0_0>(fn:execute:rescue:) ()
#20	0x00007fff5967b06b in OS_dispatch_queue.sync<τ_0_0>(execute:) ()
#21	0x000000010562b2ac in SerializedDatabase.reentrantSync<τ_0_0>(_:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/Core/SerializedDatabase.swift:153
#22	0x00000001055c8e36 in DatabaseQueue.unsafeReentrantWrite<τ_0_0>(_:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/Core/DatabaseQueue.swift:336
#23	0x00000001055c8d9d in protocol witness for DatabaseWriter.unsafeReentrantWrite<τ_0_0>(_:) in conformance DatabaseQueue ()
#24	0x000000010584c260 in ValueWriteOnlyObserver.syncStart(from:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueWriteOnlyObserver.swift:199
#25	0x000000010584b656 in ValueWriteOnlyObserver.start() at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueWriteOnlyObserver.swift:165
#26	0x00000001055ecf55 in DatabaseWriter._addWriteOnly<τ_0_0>(observation:scheduling:onChange:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/Core/DatabaseWriter.swift:379
#27	0x00000001055c9804 in DatabaseQueue._add<τ_0_0>(observation:scheduling:onChange:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/Core/DatabaseQueue.swift:360
#28	0x00000001055c9699 in protocol witness for DatabaseReader._add<τ_0_0>(observation:scheduling:onChange:) in conformance DatabaseQueue ()
#29	0x0000000105835db8 in ValueObservation.start(in:scheduling:onError:onChange:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueObservation.swift:151
#30	0x000000010583b66e in closure #1 in ValueObservation.publisher(in:scheduling:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueObservation.swift:406
#31	0x000000010583b7fc in partial apply for closure #1 in ValueObservation.publisher(in:scheduling:) ()
#32	0x000000010583c9c5 in closure #1 in DatabasePublishers.ValueSubscription.request(_:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueObservation.swift:486
#33	0x00000001055db04f in thunk for @callee_guaranteed () -> (@error @owned Error) ()
#34	0x0000000105843f54 in thunk for @callee_guaranteed () -> (@error @owned Error)partial apply ()
#35	0x0000000105822359 in NSLocking.synchronized<τ_0_0>(_:_:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/Utils/Utils.swift:183
#36	0x000000010583c3c6 in DatabasePublishers.ValueSubscription.request(_:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueObservation.swift:477
#37	0x000000010583e358 in protocol witness for Subscription.request(_:) in conformance DatabasePublishers.ValueSubscription<τ_0_0> ()
#38	0x00007fff4ed8730e in Publishers.SwitchToLatest.Outer.receive(subscription:) ()
#39	0x00007fff4ed87d60 in protocol witness for Subscriber.receive(subscription:) in conformance Publishers.SwitchToLatest<τ_0_0, τ_0_1>.Outer<τ_1_0> ()
#40	0x000000010583b9e7 in DatabasePublishers.Value.receive<τ_0_0>(subscriber:) at /Users/Shared/Garmin/prj/inReach/DerivedData/GRDBCrasher-fhpieyaxeaxlojaaergatbxzmvvx/SourcePackages/checkouts/GRDB.swift/GRDB/ValueObservation/ValueObservation.swift:432
#41	0x000000010583bb5b in protocol witness for Publisher.receive<τ_0_0>(subscriber:) in conformance DatabasePublishers.Value<τ_0_0> ()
#42	0x00007fff4eda98fa in Publishers.Map.receive<τ_0_0>(subscriber:) ()
#43	0x00007fff4ed863fe in Publishers.SwitchToLatest.receive<τ_0_0>(subscriber:) ()
#44	0x00007fff4ed126e8 in Publisher.sink(receiveCompletion:receiveValue:) ()
#45	0x0000000105547041 in AppDelegate.crash() at /Users/osmer/Desktop/GRDBCrasher/GRDBCrasher/AppDelegate.swift:45

**Comment by groue** [comment]
cc @freak4pc @nirlachman-monday

**Comment by groue** [comment]
@cosmer-work Oops, 5.22.0 has broken support for an `.immediate` observation on a database with configured with a serial `targetQueue`, as in your sample code.

**Comment by groue** [comment]
Fixed in v5.22.1!

**Comment by groue** [comment]
And again thanks for the reproducer, this was really appreciated 🙂
